feat: add support for blend=100

0x00002a opened the pull request:
<!-- Please note that we accept pull requests from anyone, but that does not mean it will be merged. -->

## What kind of change does this PR introduce?
- Feature


## Did this PR introduce a breaking change? 
- No

This adds support for `blend=100` when drawing the cursor. In terminal mode if blend=100 is in the active highlight group for `guicursor` then it makes the cursor invisible, my personal use-case for this is hiding the cursor in nerdtree and similar.

Note that I'm not sure if there is more to `blend` than just making the cursor invisible if its 100, `:help blend` says something about blend level in floating windows but I think the cursor is special, in any case I've only implemented the cursor bit.

Kethku commented on src/editor/cursor.rs:
```diff
@@ -80,6 +80,13 @@ impl Cursor {
             .unwrap_or_else(|| default_colors.foreground.unwrap())
     }
 
+    pub fn blend_is_hide(&self) -> bool {
```
maybe rename this to "should_hide"

Also rather than special casing rendering or not, why not have blend apply transparency?

Kethku replied:
![image](https://user-images.githubusercontent.com/3323631/152436734-18503390-b704-4fd2-9f87-b4e00528842f.png)
Specifically the docs indicate that this may be the intended usecase.

Kethku replied:
Actually this is a solution to the background transparency issue we've been having when trying to make the window transparent. I believe this is a great option for that

0x00002a replied:
So, just to check, do you mean apply an alpha scaled based on blend between 0 and 100 to `foreground` and `background` in `draw`? Also theres a weird issue where the cursor doesn't go invisible when opening nerdtree (which I think is a simple vsplit), even if `guicursor` is set to something with blend=100, it will update upon switching windows or entering command mode though. I'm guessing it isn't doing a redraw when it should but I don't know much about this codebase yet :p, any pointers?

Kethku replied:
Yes an alpha blend seems like the way to go. The cursor is redrawn on the base canvas with each frame, so it should be fine.

Re redraw. Thats odd, one way to force a draw is to take a reference to https://github.com/neovide/neovide/blob/main/src/redraw_scheduler.rs and then call queue_next_frame. This will force the system to draw what it's got at the next available time. I'd probably try queueing a frame when the renderer receives updated styles and or new cursor

0x00002a commented:
OK I've changed it to use transparency, so now it will scale alpha between 0 and 255 using blend between 0 and 100. I've also fixed the issue with the cursor not updating, the issue was the cursor style needs updating when the guicursor option changes. In order to do this I've added awareness of the current mode to `Editor`.

Kethku commented on src/editor/mod.rs:
```diff
@@ -79,11 +81,21 @@ impl Editor {
-                RedrawEvent::ModeInfoSet { cursor_modes } => self.mode_list = cursor_modes,
+                RedrawEvent::ModeInfoSet { cursor_modes } => {
+                    self.mode_list = cursor_modes;
+                    if let Some(curr_mode_i) = self.current_mode_index.clone() {
```
nit: please write out current, helps with somebody reading the code for the first time

Kethku commented:
Looks like there is a formatting error which can be fixed with `cargo fmt --all`. I also commented about a small naming issue, but other than that, looks great to me :) thanks so much for the change

0x00002a commented:
That should do it, my pleasure, I tried neovide out recently and I'm very much enjoying it

Kethku commented:
beautiful and probably a new speed record ;)

Will merge after the lint finishes

emmanueltouzery commented:
Note that this PR caused a regression, I opened a separate bug: https://github.com/neovide/neovide/issues/1207

emmanueltouzery commented:
Sorry for the confusion: this PR has nothing to do with the bug I opened, it didn't cause any regression, I believe. Disregard my previous message.
